File: LeetCode/2/2-Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list->
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
      ListNode* r = new ListNode(0);
      ListNode* result = r;
      while(l1 != NULL && l2 != NULL) {
        if(result->next != NULL)
          result = result->next;
        result->val = result->val + l1->val + l2->val;
        result->next = new ListNode(result->val/10);
        result->val = result->val%10;
        l1 = l1->next;
        l2 = l2->next;
      }
      while(l1 != NULL) {
        result = result->next;
        result->val = result->val + l1->val;
        result->next = new ListNode(result->val/10);
        result->val = result->val%10;
        l1 = l1->next;
      }
      while(l2 != NULL) {
        result = result->next;
        result->val = result->val + l2->val;
        result->next = new ListNode(result->val/10);
        result->val = result->val%10;
        l2 = l2->next;
      }
      if(result->next->val == 0) {
        result->next = NULL;
      }
      return r;
    }
};

File: LeetCode/2/2-Add_Two_Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  var r = new ListNode(0);
  var result = r;
  while(l1 !== null && l2 !== null) {
    if(result.next !== null)
      result = result.next;
    result.val = result.val + l1.val + l2.val;
    result.next = new ListNode(Math.floor(result.val/10));
    result.val = result.val%10;
    l1 = l1.next;
    l2 = l2.next;
  }
  while(l1 !== null) {
    result = result.next;
    result.val = result.val + l1.val;
    result.next = new ListNode(Math.floor(result.val/10));
    result.val = result.val%10;
    l1 = l1.next;
  }
  while(l2 !== null) {
    result = result.next;
    result.val = result.val + l2.val;
    result.next = new ListNode(Math.floor(result.val/10));
    result.val = result.val%10;
    l2 = l2.next;
  }
  if(result.next.val === 0) result.next = null;
  return r;
};
